File: ProjectTest/OneLang/CSharp/StdLib/PackagesFolderSource.cs
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using System;
using StdLib;
using Utils;

namespace StdLib
{
    public class PackagesFolderSource : PackageSource {
        public string packagesDir;
        
        public PackagesFolderSource(string packagesDir = "packages")
        {
            this.packagesDir = packagesDir;
        }
        
        public Task<PackageBundle> getPackageBundle(PackageId[] ids, bool cachedOnly) {
            throw new Error("Method not implemented.");
        }
        
        public async Task<PackageBundle> getAllCached() {
            var fs = ((Fs)await Global.import("fs"));
            var glob = ((Glob)await Global.import("glob"));
            var path = ((Path)await Global.import("path"));
            
            var packages = new Dictionary<string, PackageContent> {};
            var allFiles = glob.sync($"{this.packagesDir}/**/*", new Dictionary<string, bool> { ["nodir"] = true });
            foreach (var fn in allFiles) {
                var pathParts = path.relative(this.packagesDir, fn).split(new RegExp("\\/"));
                // [0]=implementations/interfaces, [1]=package-name, [2:]=path
                var type = pathParts.shift();
                var pkgDir = pathParts.shift();
                if (type != "implementations" && type != "interfaces")
                    continue;
                // skip e.g. bundle.json
                var pkgIdStr = $"{type}/{pkgDir}";
                var pkg = packages.get(pkgIdStr);
                if (pkg == null) {
                    var pkgDirParts = pkgDir.split(new RegExp("-"));
                    var version = pkgDirParts.pop().replace(new RegExp("^v"), "");
                    var pkgType = type == "implementations" ? PackageType.Implementation : PackageType.Interface;
                    var pkgId = new PackageId(pkgType, pkgDirParts.join("-"), version);
                    pkg = new PackageContent(pkgId, new Dictionary<string, string> {}, true);
                    packages.set(pkgIdStr, pkg);
                }
                pkg.files.set(pathParts.join("/"), fs.readFileSync(fn, "utf-8"));
            }
            return new PackageBundle(Object.values(packages));
        }
    }
}

File: ProjectTest/OneLang/Python/OneLang/StdLib/PackagesFolderSource.py
from OneLangStdLib import *
import re
import OneLang.StdLib.PackageManager as packMan
import OneLang._external.Native as nat

class PackagesFolderSource:
    def __init__(self, packages_dir = "packages"):
        self.packages_dir = packages_dir
    
    def get_package_bundle(self, ids, cached_only):
        raise Error("Method not implemented.")
    
    def get_all_cached(self):
        fs = import_("fs")
        glob = import_("glob")
        path = import_("path")
        
        packages = {}
        all_files = glob.sync(f'''{self.packages_dir}/**/*''', {
            "nodir": True
        })
        for fn in all_files:
            if "bundle.json" in fn:
                continue
            # TODO: hack
            path_parts = path.relative(self.packages_dir, fn).split("/")
            # [0]=implementations/interfaces, [1]=package-name, [2:]=path
            type = path_parts.pop(0)
            pkg_dir = path_parts.pop(0)
            if type != "implementations" and type != "interfaces":
                continue
            # skip e.g. bundle.json
            pkg_id_str = f'''{type}/{pkg_dir}'''
            pkg = packages.get(pkg_id_str)
            if pkg == None:
                pkg_dir_parts = pkg_dir.split("-")
                version = re.sub("^v", "", pkg_dir_parts.pop())
                pkg_type = packMan.PACKAGE_TYPE.IMPLEMENTATION if type == "implementations" else packMan.PACKAGE_TYPE.INTERFACE
                pkg_id = packMan.PackageId(pkg_type, "-".join(pkg_dir_parts), version)
                pkg = packMan.PackageContent(pkg_id, {}, True)
                packages[pkg_id_str] = pkg
            pkg.files["/".join(path_parts)] = fs.read_file_sync(fn, "utf-8")
        return packMan.PackageBundle(packages.values())